cpp/tests/utilities/column_utilities.cu

```diff
@@ -375,54 +671,104 @@ struct column_comparator {
   template <typename T>
   void operator()(column_view const& lhs,
                   column_view const& rhs,
-                  bool print_all_differences,
-                  int depth = 0)
+                  column_view const& lhs_row_indices,
+                  column_view const& rhs_row_indices,
```

Comment:
Do these need to be `column_view`s? Can't they be iterators or device_spans?

Reply:
Yeah that's what I tried initially, but it caused the compiler to spin forever.  Effectively:

```
  template<typename RowIterator>
  void operator()(column_view const& lhs,
                  column_view const& rhs,
                  RowIterator lhs_row_indices,
                  RowIterator rhs_row_indices,
                  bool print_all_differences,
                  bool throw_on_fail,
                  int depth)
  {
    ...   

    // first time in here you are passed a counting iterator.  Second time you are passed pointers
    // to plain indices.
    auto lhs_child_indices = generate_child_row_indices(lhs_l, lhs_row_indices);
    if (lhs_child_indices->size() > 0) {
      auto rhs_child         = rhs_l.get_sliced_child(rmm::cuda_stream_default);
      auto rhs_child_indices = generate_child_row_indices(rhs_l, rhs_row_indices);
      cudf::type_dispatcher(lhs_child.type(),
                            column_comparator<check_exact_equality>{},
                            lhs_child,
                            rhs_child,
                            lhs_child_indices.begin<size_type>(),
                            rhs_child_indices.begin<size_type>(),
                            print_all_differences,
                            throw_on_fail,
                            depth + 1);
    }
```

So the root calls (`expect_columns_equivalent` and `expect_columns_equal`) would just pass a counting iterator, and if you pass through a list column, they turn into arrays of indices.   I would expect the worst case here would be that it would double the amount of code generated (counting iterator and indices path for all the specializations), but the compiler hung forever.   I couldn't jiggle it into working so I went with just using indices everywhere.

Reply:
Centralizing to this thread:
https://github.com/rapidsai/cudf/pull/8666#issuecomment-881765612